File: fontend/lib/institution.ts
/**
 * Institution admin API client — the real backend behind the /admin console.
 *
 * Every call is tenant-scoped: the backend resolves the tenant from the JWT,
 * so no slug is sent. The access token comes from the tenant auth store in
 * `lib/auth.ts` (in-memory). 401s are surfaced so the shell can redirect to
 * the institution login.
 */

import { API_BASE_URL, getAccessToken, refreshAccessToken } from "./auth";
import { APIError, requestJson, guardTenantRefresh } from "./api-client";

const BASE = `${API_BASE_URL}/api/v1/institution`;

/** Re-exported so admin pages can `catch (e) { if (e instanceof …) }`. */
export { APIError as InstitutionAPIError };

const call = <T>(path: string, init: RequestInit = {}): Promise<T> =>
  requestJson<T>(
    `${BASE}${path}`,
    init,
    getAccessToken(),
    "InstitutionAPIError",
    refreshAccessToken,
    guardTenantRefresh,
  );

// ── Types ────────────────────────────────────────────────────────────────────

export interface DashboardSummary {
  tenant_id: string;
  name: string;
  slug: string;
  type: string;
  academic_year: string | null;
  counts: Record<string, number>;
  enabled_modules: string[];
  onboarding_complete: boolean;
}

export interface AcademicYear {
  id: string;
  name: string;
  start_date: string;
  end_date: string;
  is_current: boolean;
}

export interface Department {
  id: string;
  name: string;
  code: string;
  description: string | null;
  hod_id: string | null;
  hod_name: string | null;
  is_active: boolean;
  class_count: number;
  staff_count: number;
}

export interface StaffMember {
  id: string;
  name: string;
  email: string | null;
  phone: string | null;
  is_active: boolean;
  last_login_at: string | null;
  roles: string[];
  department_id: string | null;
  department_name: string | null;
}

export interface StudentRecord {
  id: string;
  name: string;
  email: string | null;
  roll_no: string | null;
  gender: string | null;
  is_active: boolean;
  enrollment: { class_name?: string; academic_year_name?: string } | null;
}

export interface ClassRecord {
  id: string;
  name: string;
  code: string;
  academic_year_name: string | null;
  department_name: string | null;
  is_active: boolean;
  grade_id: string | null;
  program_id: string | null;
  section_label: string | null;
}

/** A single section/batch within a grade or program group. */
export interface SectionRecord {
  id: string;
  name: string;
  code: string;
  section_label: string | null;
  class_teacher_id: string | null;
  class_teacher_name: string | null;
  enrolled_count: number;
  subject_count: number;
  room_no: string | null;
  is_active: boolean;
}

/** School: grade group + its sections. */
export interface ClassGradeRecord {
  id: string;
  academic_year_id: string;
  academic_year_name: string | null;
  department_id: string;
  department_name: string | null;
  name: string;
  grade_number: number;
  stream: string | null;
  is_active: boolean;
  sections: SectionRecord[];
}

/** College: program+semester group + its batches. */
export interface ClassProgramRecord {
  id: string;
  academic_year_id: string;
  academic_year_name: string | null;
  department_id: string;
  department_name: string | null;
  program_name: string;
  program_code: string;
  semester_number: number;
  is_active: boolean;
  batches: SectionRecord[];
}

export interface ModuleRow {
  key: string;
  name: string;
  is_core: boolean;
  is_enabled: boolean;
  price_monthly: number;
}

export interface SettingsInfo {
  timezone: string;
  currency: string;
  onboarding_complete: boolean;
}

export interface InstitutionProfile {
  id: string;
  name: string;
  slug: string;
  type: string;
  email: string | null;
  phone: string | null;
  address: string | null;
  city: string | null;
  state: string | null;
  country: string;
  pincode: string | null;
  website: string | null;
  logo_url: string | null;
  timezone: string;
  plan_name: string | null;
  subscription_status: string | null;
}

export interface BulkUploadRowIssue {
  row: number;
  message: string;
}

export interface BulkUploadResult {
  total: number;
  created: number;
  errors: BulkUploadRowIssue[];
  warnings: BulkUploadRowIssue[];
}

// ── Calls ────────────────────────────────────────────────────────────────────

export const fetchDashboard = () => call<DashboardSummary>("/dashboard");

export const fetchAcademicYears = () => call<AcademicYear[]>("/academic-years");
export const createAcademicYear = (payload: Partial<AcademicYear> & { name: string; start_date: string; end_date: string }) =>
  call<AcademicYear>("/academic-years", { method: "POST", body: JSON.stringify(payload) });
export const updateAcademicYear = (id: string, payload: Partial<AcademicYear>) =>
  call<AcademicYear>(`/academic-years/${id}`, { method: "PUT", body: JSON.stringify(payload) });
export const deleteAcademicYear = (id: string) =>
  call<null>(`/academic-years/${id}`, { method: "DELETE" });

export const fetchDepartments = () => call<Department[]>("/departments");
export const createDepartment = (payload: { name: string; code: string; description?: string }) =>
  call<Department>("/departments", { method: "POST", body: JSON.stringify(payload) });
export const updateDepartment = (id: string, payload: { hod_id?: string | null; name?: string; description?: string; is_active?: boolean }) =>
  call<Department>(`/departments/${id}`, { method: "PUT", body: JSON.stringify(payload) });

export const fetchStaff = () => call<StaffMember[]>("/staff");
export const uploadStaff = (file: File) => {
  const form = new FormData();
  form.append("file", file);
  return call<BulkUploadResult>("/staff/bulk", { method: "POST", body: form });
};
export const inviteStaff = (payload: {
  name: string;
  email: string;
  phone?: string;
  role: string;
  departmentId?: string;
}) =>
  call<StaffMember>("/staff", {
    method: "POST",
    body: JSON.stringify({
      name: payload.name,
      email: payload.email,
      phone: payload.phone,
      role: payload.role,
      department_id: payload.departmentId,
    }),
  });
export const assignStaffRole = (userId: string, roleName: string, departmentId?: string) =>
  call<StaffMember>(`/staff/${userId}/roles`, {
    method: "PUT",
    body: JSON.stringify({ role_name: roleName, department_id: departmentId }),
  });
export const revokeStaffRole = (userId: string, roleName: string, departmentId?: string) =>
  call<StaffMember>(
    `/staff/${userId}/roles/${encodeURIComponent(roleName)}${
      departmentId ? `?department_id=${encodeURIComponent(departmentId)}` : ""
    }`,
    { method: "DELETE" },
  );
export const updateStaff = (
  userId: string,
  payload: { name?: string; email?: string; phone?: string; departmentId?: string }
) =>
  call<StaffMember>(`/staff/${userId}`, {
    method: "PUT",
    body: JSON.stringify({
      name: payload.name,
      email: payload.email,
      phone: payload.phone,
      department_id: payload.departmentId,
    }),
  });
export const deleteStaff = (userId: string) =>
  call<null>(`/staff/${userId}`, { method: "DELETE" });
export const setStaffActive = (userId: string, active: boolean) =>
  call<StaffMember>(`/staff/${userId}/active?active=${active}`, { method: "PUT" });

export const fetchStudents = () => call<StudentRecord[]>("/students");
export const fetchClasses = () => call<ClassRecord[]>("/classes");
export const createStudent = (payload: { name: string; roll_no: string; email?: string; gender?: string; class_id?: string }) =>
  call<StudentRecord>("/students", { method: "POST", body: JSON.stringify(payload) });

// ── School grade wizard ─────────────────────────────────────────────────────

export const fetchGrades = (academic_year_id?: string) =>
  call<ClassGradeRecord[]>(`/grades${academic_year_id ? `?academic_year_id=${academic_year_id}` : ""}`);

export const createGrade = (payload: {
  academic_year_id: string;
  department_id: string;
  grade_number: number;
  stream?: string;
  sections: string[];
  max_strength?: number;
  class_teacher_id?: string;
}) => call<ClassGradeRecord>("/grades", { method: "POST", body: JSON.stringify(payload) });

export const deleteGrade = (grade_id: string) =>
  call<null>(`/grades/${grade_id}`, { method: "DELETE" });

// ── College program wizard ──────────────────────────────────────────────────

export const fetchPrograms = (department_id?: string, academic_year_id?: string) => {
  const params = new URLSearchParams();
  if (department_id) params.set("department_id", department_id);
  if (academic_year_id) params.set("academic_year_id", academic_year_id);
  const qs = params.toString();
  return call<ClassProgramRecord[]>(`/programs${qs ? `?${qs}` : ""}`);
};

export const createProgram = (payload: {
  academic_year_id: string;
  department_id: string;
  program_name: string;
  program_code: string;
  semester_number: number;
  batches: string[];
  max_strength?: number;
  class_teacher_id?: string;
}) => call<ClassProgramRecord>("/programs", { method: "POST", body: JSON.stringify(payload) });

export const deleteProgram = (program_id: string) =>
  call<null>(`/programs/${program_id}`, { method: "DELETE" });

export const fetchModules = () => call<ModuleRow[]>("/modules");
export const toggleModule = (key: string, enabled: boolean) =>
  call<ModuleRow>(`/modules/${key}`, { method: "PUT", body: JSON.stringify({ enabled }) });

export const fetchSettings = () => call<SettingsInfo>("/settings");
export const updateSettings = (payload: Partial<SettingsInfo>) =>
  call<SettingsInfo>("/settings", { method: "PUT", body: JSON.stringify(payload) });

export const fetchProfile = () => call<InstitutionProfile>("/profile");
export const updateProfile = (payload: Partial<InstitutionProfile>) =>
  call<InstitutionProfile>("/profile", { method: "PUT", body: JSON.stringify(payload) });

File: backend/app/schemas/institution.py
"""
Pydantic Schemas — institution admin management API

DTOs for the day-to-day admin flows: dashboard, academic structure (years,
departments, classes, subjects), people (staff/users, students, enrollments),
modules, settings and the institution profile. All scoped to the admin's
tenant via the JWT.
"""

import uuid
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, EmailStr, Field

from app.schemas.common import APIResponse

# ── Dashboard ────────────────────────────────────────────────────────────────

class DashboardSummary(BaseModel):
    tenant_id: uuid.UUID
    name: str
    slug: str
    type: str
    academic_year: str | None = None
    counts: dict[str, int]
    enabled_modules: list[str]
    onboarding_complete: bool


# ── Academic years ───────────────────────────────────────────────────────────

class AcademicYearCreate(BaseModel):
    name: str = Field(..., max_length=50)
    start_date: date
    end_date: date
    is_current: bool = False


class AcademicYearUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=50)
    start_date: date | None = None
    end_date: date | None = None
    is_current: bool | None = None


class AcademicYearOut(BaseModel):
    id: uuid.UUID
    name: str
    start_date: date
    end_date: date
    is_current: bool


# ── Departments ──────────────────────────────────────────────────────────────

class DepartmentCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=255)
    code: str = Field(..., min_length=1, max_length=20)
    description: str | None = None
    hod_id: uuid.UUID | None = None


class DepartmentUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=255)
    description: str | None = None
    hod_id: uuid.UUID | None = None
    is_active: bool | None = None


class DepartmentOut(BaseModel):
    id: uuid.UUID
    name: str
    code: str
    description: str | None = None
    hod_id: uuid.UUID | None = None
    hod_name: str | None = None
    is_active: bool
    class_count: int = 0
    staff_count: int = 0


# ── Class grades (school) + programs (college) ─────────────────────────────

# School: one ClassGrade row groups the N sections of e.g. "Class 11 Science"
class ClassGradeCreate(BaseModel):
    """Wizard payload: create a school grade group + one section per letter."""
    academic_year_id: uuid.UUID
    # Section lives inside a department even for school (for FK integrity)
    department_id: uuid.UUID
    grade_number: int = Field(..., ge=1, le=12, description="Grade 1–12")
    stream: str | None = Field(default=None, max_length=50, description="Science/Commerce/Arts or custom")
    # Section labels: ["A","B","C"] — each becomes one SchoolClass (Academic Group)
    sections: list[str] = Field(..., min_length=1, description="Section letters/labels, e.g. ['A','B','C']")
    max_strength: int = Field(default=60, ge=1)
    # Optional: assign the same class teacher to all sections, or leave empty
    # (per-section assignment can be done from the class detail page)
    class_teacher_id: uuid.UUID | None = None


class SectionOut(BaseModel):
    """A single academic group (section) within a grade or program."""
    id: uuid.UUID
    name: str
    code: str
    section_label: str | None = None
    class_teacher_id: uuid.UUID | None = None
    class_teacher_name: str | None = None
    enrolled_count: int = 0
    subject_count: int = 0
    room_no: str | None = None
    is_active: bool


class ClassGradeOut(BaseModel):
    id: uuid.UUID
    academic_year_id: uuid.UUID
    academic_year_name: str | None = None
    department_id: uuid.UUID
    department_name: str | None = None
    name: str              # "Class 11"
    grade_number: int
    stream: str | None     # "Science" / None
    is_active: bool
    sections: list[SectionOut] = Field(default_factory=list)


# College: one ClassProgram row groups the N batches of e.g. "B.Tech CSE Semester 3"
class ClassProgramCreate(BaseModel):
    """Wizard payload: create a college program+semester + one batch per label."""
    academic_year_id: uuid.UUID
    department_id: uuid.UUID
    program_name: str = Field(..., min_length=2, max_length=200, description="B.Tech CSE")
    program_code: str = Field(..., min_length=1, max_length=30, description="BTCSE")
    semester_number: int = Field(..., ge=1, description="Semester number")
    # Batch labels: ["A","B"] — each becomes one SchoolClass (Academic Group)
    batches: list[str] = Field(..., min_length=1, description="Batch labels, e.g. ['A','B']")
    max_strength: int = Field(default=60, ge=1)
    class_teacher_id: uuid.UUID | None = None


class ClassProgramOut(BaseModel):
    id: uuid.UUID
    academic_year_id: uuid.UUID
    academic_year_name: str | None = None
    department_id: uuid.UUID
    department_name: str | None = None
    program_name: str      # "B.Tech CSE"
    program_code: str      # "BTCSE"
    semester_number: int
    is_active: bool
    batches: list[SectionOut] = Field(default_factory=list)


# ── Classes (Academic Groups) ───────────────────────────────────────────

class ClassCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    code: str = Field(..., min_length=1, max_length=20)
    department_id: uuid.UUID
    academic_year_id: uuid.UUID
    max_strength: int = Field(default=60, ge=1)
    room_no: str | None = Field(default=None, max_length=20)
    class_teacher_id: uuid.UUID | None = None


class ClassUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=100)
    max_strength: int | None = Field(default=None, ge=1)
    room_no: str | None = None
    class_teacher_id: uuid.UUID | None = None
    is_active: bool | None = None


class ClassOut(BaseModel):
    id: uuid.UUID
    name: str
    code: str
    department_id: uuid.UUID
    department_name: str | None = None
    academic_year_id: uuid.UUID
    academic_year_name: str | None = None
    max_strength: int
    room_no: str | None = None
    class_teacher_id: uuid.UUID | None = None
    class_teacher_name: str | None = None
    is_active: bool
    enrolled_count: int = 0
    subject_count: int = 0
    # ── Hierarchy parent fields (None for flat / legacy classes) ───────────
    grade_id: uuid.UUID | None = None
    program_id: uuid.UUID | None = None
    section_label: str | None = None


# ── Subjects ─────────────────────────────────────────────────────────────────

class SubjectCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    code: str = Field(..., min_length=1, max_length=30)
    class_id: uuid.UUID
    subject_type: str = Field(default="THEORY", pattern="^(THEORY|PRACTICAL|ELECTIVE|PROJECT)$")
    credits: int | None = Field(default=None, ge=0)
    max_marks: int = Field(default=100, ge=1)
    passing_marks: int = Field(default=35, ge=0)


class SubjectUpdate(BaseModel):
    name: str | None = None
    subject_type: str | None = Field(default=None, pattern="^(THEORY|PRACTICAL|ELECTIVE|PROJECT)$")
    credits: int | None = Field(default=None, ge=0)
    max_marks: int | None = Field(default=None, ge=1)
    passing_marks: int | None = Field(default=None, ge=0)
    is_active: bool | None = None


class SubjectOut(BaseModel):
    id: uuid.UUID
    name: str
    code: str
    class_id: uuid.UUID
    class_name: str | None = None
    subject_type: str
    credits: int | None = None
    max_marks: int
    passing_marks: int
    is_active: bool
    teachers: list[dict[str, Any]] = Field(default_factory=list)


# ── People: staff / users ────────────────────────────────────────────────────

class StaffInvite(BaseModel):
    name: str = Field(..., min_length=2, max_length=255)
    email: EmailStr
    phone: str | None = Field(default=None, max_length=20)
    role: str = Field(..., max_length=50, description="Role name, e.g. TEACHER")
    department_id: uuid.UUID | None = None


class StaffUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=255)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=20)
    department_id: uuid.UUID | None = None



class RoleAssignmentOut(BaseModel):
    role_id: uuid.UUID
    role_name: str
    is_active: bool


class StaffOut(BaseModel):
    id: uuid.UUID
    name: str
    email: str | None = None
    phone: str | None = None
    is_active: bool
    last_login_at: datetime | None = None
    roles: list[str] = Field(default_factory=list)
    department_id: uuid.UUID | None = None
    department_name: str | None = None


class AssignRoleRequest(BaseModel):
    role_name: str = Field(..., max_length=50)
    # Required when assigning VICE_PRINCIPAL: it is the delegated department
    # scope, not an optional profile field.
    department_id: uuid.UUID | None = None


# ── People: students + enrollments ───────────────────────────────────────────

class StudentCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=255)
    email: EmailStr | None = None
    roll_no: str = Field(..., min_length=1, max_length=50)
    gender: str | None = Field(default=None, pattern="^(MALE|FEMALE|OTHER)$")
    date_of_birth: date | None = None
    class_id: uuid.UUID | None = None


class StudentOut(BaseModel):
    id: uuid.UUID
    name: str
    email: str | None = None
    roll_no: str | None = None
    gender: str | None = None
    is_active: bool
    enrollment: dict[str, Any] | None = None


class BulkUploadRowIssue(BaseModel):
    """One problem found in a CSV row during bulk student import."""
    row: int
    message: str


class BulkUploadResult(BaseModel):
    total: int
    created: int
    errors: list[BulkUploadRowIssue] = Field(default_factory=list)
    warnings: list[BulkUploadRowIssue] = Field(default_factory=list)


class EnrollmentCreate(BaseModel):
    student_id: uuid.UUID
    class_id: uuid.UUID
    academic_year_id: uuid.UUID | None = None
    roll_number: str | None = Field(default=None, max_length=50)


class EnrollmentOut(BaseModel):
    id: uuid.UUID
    student_id: uuid.UUID
    student_name: str
    class_id: uuid.UUID
    class_name: str
    academic_year_id: uuid.UUID
    academic_year_name: str
    roll_number: str | None = None
    status: str
    enrollment_date: date


# ── Modules + settings ───────────────────────────────────────────────────────

class ModuleOut(BaseModel):
    key: str
    name: str
    is_core: bool
    is_enabled: bool
    price_monthly: float = 0


class ModuleToggle(BaseModel):
    enabled: bool


class SettingsOut(BaseModel):
    timezone: str
    currency: str
    onboarding_complete: bool


class SettingsUpdate(BaseModel):
    timezone: str | None = None
    currency: str | None = None


# ── Institution profile ──────────────────────────────────────────────────────

class InstitutionProfileOut(BaseModel):
    id: uuid.UUID
    name: str
    slug: str
    type: str
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    city: str | None = None
    state: str | None = None
    country: str
    pincode: str | None = None
    website: str | None = None
    logo_url: str | None = None
    timezone: str
    plan_name: str | None = None
    subscription_status: str | None = None


class InstitutionProfileUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=20)
    address: str | None = None
    city: str | None = Field(default=None, max_length=100)
    state: str | None = Field(default=None, max_length=100)
    pincode: str | None = Field(default=None, max_length=20)
    website: str | None = Field(default=None, max_length=255)
    logo_url: str | None = None


# ── Typed APIResponse aliases ────────────────────────────────────────────────

APIResponseDashboard = APIResponse[DashboardSummary]
APIResponseYears = APIResponse[list[AcademicYearOut]]
APIResponseYear = APIResponse[AcademicYearOut]
APIResponseDepartments = APIResponse[list[DepartmentOut]]
APIResponseDepartment = APIResponse[DepartmentOut]
APIResponseGrades = APIResponse[list[ClassGradeOut]]
APIResponseGrade = APIResponse[ClassGradeOut]
APIResponsePrograms = APIResponse[list[ClassProgramOut]]
APIResponseProgram = APIResponse[ClassProgramOut]
APIResponseClasses = APIResponse[list[ClassOut]]
APIResponseClass = APIResponse[ClassOut]
APIResponseSubjects = APIResponse[list[SubjectOut]]
APIResponseSubject = APIResponse[SubjectOut]
APIResponseStaff = APIResponse[list[StaffOut]]
APIResponseStaffOne = APIResponse[StaffOut]
APIResponseStudents = APIResponse[list[StudentOut]]
APIResponseStudent = APIResponse[StudentOut]
APIResponseBulk = APIResponse[BulkUploadResult]
APIResponseEnrollments = APIResponse[list[EnrollmentOut]]
APIResponseEnrollment = APIResponse[EnrollmentOut]
APIResponseModules = APIResponse[list[ModuleOut]]
APIResponseModule = APIResponse[ModuleOut]
APIResponseSettings = APIResponse[SettingsOut]
APIResponseProfile = APIResponse[InstitutionProfileOut]
